apps/evm_family/evm_sign_msg.c

```diff
@@ -0,0 +1,447 @@
+ * GLOBAL VARIABLES
+ *****************************************************************************/
+
+/*****************************************************************************
+ * STATIC FUNCTIONS
+ *****************************************************************************/
+
+static bool check_which_request(const evm_query_t *query,
+                                pb_size_t which_request) {
+  if (which_request != query->get_public_keys.which_request) {
+    evm_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                   ERROR_DATA_FLOW_INVALID_REQUEST);
+    return false;
+  }
+
+  return true;
+}
+
+static bool validate_initiate_query(evm_sign_msg_initiate_request_t *init_req) {
+  switch (init_req->message_type) {
+    case EVM_SIGN_MSG_TYPE_ETH_SIGN:
+    case EVM_SIGN_MSG_TYPE_PERSONAL_SIGN:
+      // TODO: Add evm typed data struct message signing
+      // case EVM_SIGN_MSG_TYPE_SIGN_TYPED_DATA:
+
+      // TODO before PR: Verify derivation path count
+      if (!evm_derivation_path_guard(init_req->derivation_path,
+                                     init_req->derivation_path_count)) {
+        break;
+      }
+      if (MAX_MSG_DATA_SIZE < init_req->total_msg_size) {
+        evm_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                       ERROR_DATA_FLOW_INVALID_DATA);
+        break;
+      }
+      return true;
+      break;
+
+    default:
+      break;
+  }
+
+  evm_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                 ERROR_DATA_FLOW_INVALID_DATA);
+  return false;
+}
+
+static bool handle_initiate_query(evm_query_t *query) {
+  char msg[100] = "";
+  uint8_t wallet_name[NAME_SIZE] = {0};
+  evm_result_t response = init_evm_result(EVM_RESULT_SIGN_MSG_TAG);
+  if (!check_which_request(query, EVM_SIGN_MSG_REQUEST_INITIATE_TAG) ||
+      !validate_initiate_query(&query->sign_msg.initiate) ||
+      !get_wallet_name_by_id(query->sign_msg.initiate.wallet_id,
+                             (uint8_t *)wallet_name,
+                             evm_send_error)) {
+    return false;
+  }
+
+  snprintf(msg,
+           sizeof(msg),
+           UI_TEXT_SIGN_PROMPT,
+           g_evm_app->lunit_name,
+           g_evm_app->name,
+           wallet_name);
+  // Take user consent to sign message for the wallet
+  if (!core_confirmation(msg, evm_send_error)) {
+    return false;
+  }
+
+  set_app_flow_status(EVM_SIGN_MSG_STATUS_CONFIRM);
+
+  memcpy(&(sign_msg_ctx.init),
+         &(query->sign_msg.initiate),
+         sizeof(query->sign_msg.initiate));
+
+  response.sign_msg.which_response = EVM_SIGN_MSG_RESPONSE_CONFIRMATION_TAG;
+  response.sign_msg.confirmation.dummy_field = 0;
+  evm_send_result(&response);
+
+  return true;
+}
+
+static bool get_msg_data(evm_query_t *query) {
+  bool result = false;
+  evm_result_t response = init_evm_result(EVM_RESULT_SIGN_MSG_TAG);
+  common_chunk_payload_t *chunk_payload = NULL;
+  uint32_t offset = 0;
+
+  sign_msg_ctx.msg_data = NULL;
+
+  do {
```

Comment:
Unify stitching logic with eth sign txn

Reply:
Updated and tested, there's change required in evm sign txn logic for sending a chunk ack